File: tam_agent/analysis/top_down.py
"""Top-down TAM estimation."""

from __future__ import annotations

from tam_agent.llm.client import LLMClient
from tam_agent.llm.prompts import SYSTEM_PROMPT, TOP_DOWN_TAM
from tam_agent.models import (
    CompanyInput,
    IndustryClassification,
    MarketDataPoint,
    MarketEstimate,
    SearchResult,
)
from tam_agent.research.data_extraction import extract_market_data, format_search_data_section
from tam_agent.research.web_search import WebSearcher


def estimate_top_down(
    llm: LLMClient,
    searcher: WebSearcher,
    company: CompanyInput,
    industry: IndustryClassification,
) -> tuple[MarketEstimate, list[SearchResult], list[MarketDataPoint]]:
    """Estimate TAM using top-down approach (industry reports -> narrow down)."""

    # Generate search queries
    context = (
        f"Company: {company.name}, Industry: {company.industry} "
        f"(NAICS {industry.naics_code}: {industry.naics_description}), "
        f"Geography: {company.geography}, Sub-segments: {', '.join(industry.sub_segments)}"
    )
    queries = llm.generate_search_queries(
        f"Find total market size data for top-down TAM estimation.\n{context}",
        num_queries=4,
    )

    # Search and extract data
    search_results = searcher.multi_search(queries)
    data_points = extract_market_data(llm, search_results, context)
    search_data_section = format_search_data_section(search_results, data_points)

    # Estimate
    prompt = TOP_DOWN_TAM.format(
        name=company.name,
        industry=company.industry,
        naics_code=industry.naics_code,
        naics_description=industry.naics_description,
        geography=company.geography,
        sub_segments=", ".join(industry.sub_segments),
        search_data_section=search_data_section,
    )

    estimate = llm.structured_query(
        prompt=prompt,
        response_model=MarketEstimate,
        system=SYSTEM_PROMPT,
    )
    return estimate, search_results, data_points

File: src/analysis/top-down.ts
/** Top-down TAM estimation. */

import type { LLMClient } from "../llm/client.js";
import { topDownTamPrompt, SYSTEM_PROMPT } from "../llm/prompts.js";
import {
  MarketEstimateSchema,
  type CompanyInput,
  type IndustryClassification,
  type MarketDataPoint,
  type MarketEstimate,
  type SearchResult,
} from "../models.js";
import { extractMarketData, formatSearchDataSection } from "../research/data-extraction.js";
import type { WebSearcher } from "../research/web-search.js";

export async function estimateTopDown(
  llm: LLMClient,
  searcher: WebSearcher,
  company: CompanyInput,
  industry: IndustryClassification,
): Promise<{
  estimate: MarketEstimate;
  searchResults: SearchResult[];
  dataPoints: MarketDataPoint[];
}> {
  const context =
    `Company: ${company.name}, Industry: ${company.industry} ` +
    `(NAICS ${industry.naics_code}: ${industry.naics_description}), ` +
    `Geography: ${company.geography}, Sub-segments: ${industry.sub_segments.join(", ")}`;

  const queries = await llm.generateSearchQueries(
    `Find total market size data for top-down TAM estimation.\n${context}`,
    4,
  );

  const searchResults = await searcher.multiSearch(queries);
  const dataPoints = await extractMarketData(llm, searchResults, context);
  const searchDataSection = formatSearchDataSection(searchResults, dataPoints);

  const estimate = await llm.structuredQuery(
    topDownTamPrompt({
      name: company.name,
      industry: company.industry,
      naics_code: industry.naics_code,
      naics_description: industry.naics_description,
      geography: company.geography,
      sub_segments: industry.sub_segments.join(", "),
      search_data_section: searchDataSection,
    }),
    MarketEstimateSchema,
    "MarketEstimate",
    SYSTEM_PROMPT,
  );

  return { estimate, searchResults, dataPoints };
}
